flaky testHTTPUpgradeTestCase.testUpgradeWithUpgradePayloadInlineWithRequestWorks (in docker)

### Expected behavior
all tests should pass

### Actual behavior
The following test case fails
```bash
Test Case 'HTTPUpgradeTestCase.testUpgradeWithUpgradePayloadInlineWithRequestWorks' started at 2018-05-23 23:43:59.654
/code/Tests/NIOHTTP1Tests/HTTPUpgradeTests.swift:850: error: HTTPUpgradeTestCase.testUpgradeWithUpgradePayloadInlineWithRequestWorks : XCTAssertEqual failed: ("1") is not equal to ("2") -
/code/Tests/NIOHTTP1Tests/HTTPUpgradeTests.swift:854: error: HTTPUpgradeTestCase.testUpgradeWithUpgradePayloadInlineWithRequestWorks : XCTAssertEqual failed: ("Optional("A")") is not equal to ("Optional("AB")") -
```

### Steps to reproduce

1.  ```git clone https://github.com/apple/swift-nio.git```
2. ``` docker-compose -f docker/docker-compose.yaml -f docker/docker-compose.1604.41.yaml run test```

### Software

MacOS High Sierra 10.13.4 (17E202)
Docker Version 18.04.0-ce-mac62 (23965)
Channel: edge
Compose: 1.21.0


### SwiftNIO version/commit hash

398b950eeedb8215bfdafdc07ebf92a7e0722980

### Swift & OS version (output of `swift --version && uname -a`)
Apple Swift version 4.1 (swift-4.1-RELEASE)
Target: x86_64-apple-darwin17.5.0
Darwin Karims-MacBook-Pro.local 17.5.0 Darwin Kernel Version 17.5.0: Fri Apr 13 19:32:32 PDT 2018; root:xnu-4570.51.2~1/RELEASE_X86_64 x86_64

--- a/Tests/NIOHTTP1Tests/HTTPUpgradeTests.swift
+++ b/Tests/NIOHTTP1Tests/HTTPUpgradeTests.swift
@@ -816,6 +816,7 @@ class HTTPUpgradeTestCase: XCTestCase {
     }
     
     func testUpgradeWithUpgradePayloadInlineWithRequestWorks() throws {
+        enum ReceivedTheWrongThingError: Error { case error }
         var upgradeRequest: HTTPRequestHead? = nil
         var upgradeHandlerCbFired = false
         var upgraderCbFired = false
@@ -833,10 +834,16 @@ class HTTPUpgradeTestCase: XCTestCase {
                 case closed
             }
             
+            private let firstByteDonePromise: EventLoopPromise<Void>
+            private let secondByteDonePromise: EventLoopPromise<Void>
             private let allDonePromise: EventLoopPromise<Void>
             private var state = State.fresh
             
-            init(allDonePromise: EventLoopPromise<Void>) {
+            init(firstByteDonePromise: EventLoopPromise<Void>,
+                 secondByteDonePromise: EventLoopPromise<Void>,
+                 allDonePromise: EventLoopPromise<Void>) {
+                self.firstByteDonePromise = firstByteDonePromise
+                self.secondByteDonePromise = secondByteDonePromise
                 self.allDonePromise = allDonePromise
             }
             
@@ -853,10 +860,20 @@ class HTTPUpgradeTestCase: XCTestCase {
                 case .added:
                     XCTAssertEqual("A", stringRead)
                     self.state = .inlineDataRead
+                    if stringRead == .some("A") {
+                        self.firstByteDonePromise.succeed(result: ())
+                    } else {
+                        self.firstByteDonePromise.fail(error: ReceivedTheWrongThingError.error)
+                    }
                 case .inlineDataRead:
                     XCTAssertEqual("B", stringRead)
                     self.state = .extraDataRead
                     ctx.channel.close(promise: nil)
+                    if stringRead == .some("B") {
+                        self.secondByteDonePromise.succeed(result: ())
+                    } else {
+                        self.secondByteDonePromise.fail(error: ReceivedTheWrongThingError.error)
+                    }
                 default:
                     XCTFail("channel read in wrong state \(self.state)")
                 }
@@ -867,7 +884,7 @@ class HTTPUpgradeTestCase: XCTestCase {
                 self.state = .closed
                 ctx.close(mode: mode, promise: promise)
                 
-                allDonePromise.succeed(result: ())
+                self.allDonePromise.succeed(result: ())
             }
         }
         
@@ -877,6 +894,8 @@ class HTTPUpgradeTestCase: XCTestCase {
             upgraderCbFired = true
         }
         
+        let firstByteDonePromise: EventLoopPromise<Void> = EmbeddedEventLoop().newPromise()
+        let secondByteDonePromise: EventLoopPromise<Void> = EmbeddedEventLoop().newPromise()
         let allDonePromise: EventLoopPromise<Void> = EmbeddedEventLoop().newPromise()
         let (group, server, client, connectedServer) = try setUpTestWithAutoremoval(upgraders: [upgrader],
                                                                                     extraHandlers: []) { (ctx) in
@@ -884,7 +903,9 @@ class HTTPUpgradeTestCase: XCTestCase {
                                                                                         XCTAssertNil(upgradeRequest)
                                                                                         upgradeHandlerCbFired = true
                                                                                         
-                                                                                        _ = ctx.channel.pipeline.add(handler: CheckWeReadInlineAndExtraData(allDonePromise: allDonePromise))
+                                                                                        _ = ctx.channel.pipeline.add(handler: CheckWeReadInlineAndExtraData(firstByteDonePromise: firstByteDonePromise,
+                                                                                                                                                            secondByteDonePromise: secondByteDonePromise,
+                                                                                                                                                            allDonePromise: allDonePromise))
         }
         defer {
             XCTAssertNoThrow(try group.syncShutdownGracefully())
@@ -905,8 +926,14 @@ class HTTPUpgradeTestCase: XCTestCase {
         request += "A"
         XCTAssertNoThrow(try client.writeAndFlush(NIOAny(ByteBuffer.forString(request))).wait())
 
+        XCTAssertNoThrow(try firstByteDonePromise.futureResult.wait() as Void)
+
         XCTAssertNoThrow(try client.writeAndFlush(NIOAny(ByteBuffer.forString("B"))).wait())
         
+        XCTAssertNoThrow(try secondByteDonePromise.futureResult.wait() as Void)
+
+        XCTAssertNoThrow(try allDonePromise.futureResult.wait() as Void)
+
         // Let the machinery do its thing.
         XCTAssertNoThrow(try completePromise.futureResult.wait())
         

